### ISSUE TITLE
Reproducing BigCodeBench Results — Seeking Guidance

### ISSUE THREAD

**Issue by Lu-Yang666** [MemTensor/MemRL#12]
Hi, thank you for the great work on MemRL! I'm trying to reproduce the BigCodeBench results and would appreciate some guidance on a few points.

### Environment
- Code: latest main branch from the GitHub repo
- Python 3.10, running in a Singularity container
- Qdrant for vector storage (as configured in the repo)

### Model & API
- **LLM**: `gpt-4o-2024-11-20` via an OpenAI-compatible API proxy (not the official OpenAI endpoint directly)
- **Embedding**: `text-embedding-3-large` via the same proxy
- **Temperature**: 0.0

### Data Split
- BigCodeBench-Instruct (Full), 1140 tasks, 7:3 train/val split with **seed 42** → 798 train + 342 val
- 90 tasks (48 train + 42 val) are consistently blocked by our API's content filter. These tasks are skipped every run and never enter the memory store.

### Hyperparameters (following Table 8)

| Parameter | My Setting | Paper (Table 8) |
|-----------|-----------|-----------------|
| α (learning rate) | 0.3 | 0.3 |
| λ (Q-weight balance) | 0.5 | 0.5 |
| δ (sim_threshold) | 0.38 | 0.38 |
| k1 (Phase-A recall) | 10 | 10 |
| k2 (Phase-B select) | 5 | 5 |
| Qinit | 0.0 | 0.0 |
| build_strategy | proceduralization | proceduralization |
| retrieve_strategy | query | query |
| update_strategy | adjustment | adjustment |
| sim_norm_mean | 0.186 | ? |
| sim_norm_std | 0.094 | ? |

## My Results

Excluding the 90 content-filtered tasks, evaluated on 750 train + 300 val:

| Epoch | Train SR | Val SR |
|-------|----------|--------|
| E1 | 44.5% | 45.7% |
| E2 | 47.2% | 44.7% |
| E3 | 48.8% | 45.7% |
| E4 | 49.1% | 47.0% |
| E5 | 49.3% | 44.3% |
| E6 | 50.1% | 44.3% |
| E7 | 50.7% | 44.7% |

**No-Memory Baseline** (retrieve_k=0, 1 epoch): Train 44.7%, Val 46.0%

Paper reported: No-Memory 48.5%, MemRL Runtime Learning 59.5%, Transfer Learning 50.8%.

## Questions

1. Could you share which exact GPT-4o API version/snapshot was used in the experiments? My baseline (46.0%) is about 2.5pp lower than the paper's (48.5%), and I suspect this is due to model/API differences.

2. What do Table 1 "Last Epoch Success Rate" and Table 2 "Transfer Learning Success Rate" correspond to exactly — is Table 1 the train set SR at epoch 10, and Table 2 the best val SR across epochs?

3. What `sim_norm_mean` and `sim_norm_std` values were used for BigCodeBench? The paper mentions δ is set at the "top 20% quantile of pairwise cosine similarity." My computed values (0.186, 0.094) differ from defaults in the repo (0.31, 0.10), and I'm not sure which is correct.

4. I observe train SR improving about 1pp per epoch (44.5% → 50.7% over 7 epochs), while val SR remains flat around 44-47%. Is this training-set improvement rate consistent with your experiments, or should I expect faster convergence?

5. In your experiments, did MemRL also improve at roughly ~1pp per epoch on the training set over 10 epochs (e.g., from ~48.5% to ~59.5%), or was the improvement curve different (faster early on, etc.)? 

Thank you for your time! Happy to provide more details about our setup if helpful!

**Comment by ai-jiaqian** [comment]
Hi @Lu-Yang666, thanks for the very detailed setup description — this kind of careful reporting makes diagnosis much easier. Let me address your five questions, but first I'd like to ask you to verify two things on your side, because they may already explain a large part of the gap.

## ⚠️ Before everything: please confirm two things

### 1. Are you on a code version after **2026-04-09**?

On **2026-04-09** we landed a critical BCB-specific bugfix (commit [`dc4d50a`](https://github.com/MemTensor/MemRL/commit/dc4d50a)):

> 🐛 **fix(bcb): make `memory_budget_tokens` configurable, default unlimited**
>
> Previously, the BCB memory context was hardcoded to **2000 characters total**, split across the 5 retrieved memories (~400 chars each). This severely truncated proceduralized memories and made the memory injection nearly useless on BCB.

Since you filed this issue on 2026-05-08, your "latest main" *should* include it — but please verify:

```bash
# (a) confirm the commit is in your checkout
git log --oneline dc4d50a -1
# Expected: dc4d50a 🐛 fix(bcb): make memory_budget_tokens configurable, default unlimited

# (b) confirm the config value
grep memory_budget_tokens configs/rl_bcb_config.yaml
# Expected:   memory_budget_tokens: 0
```

Also please confirm `memory_budget_tokens` prints as `0` in your runtime config dump / log, not `2000`. If you see `2000` anywhere, that alone could explain a large part of the Runtime Learning gap, since proceduralized memory is much longer than 400 chars.

### 2. If the code version is fine, the 90 content-filtered tasks become the next suspect

If item 1 checks out, then the next thing worth flagging is the 90 tasks (48 train + 42 val) being consistently filtered by your proxy. This has two distinct effects beyond just the denominator change:

1. **Memory learning signal is reduced**: 48 train tasks not entering the memory store means MemRL gets ~6% fewer add-memory / Q-update opportunities per epoch. This directly slows down the rate at which useful procedural memories can be discovered and reinforced.
2. **The skipped tasks are likely non-random**: content filters typically reject tasks involving subprocess, network requests, credential/email handling, etc. The remaining 1050 may be skewed in either direction vs. the paper's 1140-task numbers.

This won't make the gap go away by itself, but it should be kept in mind as a structural factor when comparing your numbers against the paper's.

---

## Answers to your five questions

### Q1 — GPT-4o snapshot version

Honest answer: I can't authoritatively confirm the exact snapshot. We ran our BCB experiments through an OpenAI-compatible **proxy** where the model is exposed simply as `gpt-4o` (the alias), not a dated snapshot. The underlying snapshot the proxy routed to at the time of our experiments isn't something we can verify retroactively.

For reference, the BigCodeBench leaderboard reports calibrated Pass@1 on BCB-Instruct (Full):

| Snapshot | BCB-Instruct (Full) |
|---|---|
| `gpt-4o-2024-05-13` | 51.1 |
| `gpt-4o-2024-08-06` | (not submitted) |
| `gpt-4o-2024-11-20` | 48.0 |

(Source: [bigcode/bigcodebench-results](https://huggingface.co/datasets/bigcode/bigcodebench-results))

Our paper's No-Memory baseline (48.5) is closest to the `2024-11-20` leaderboard number (48.0). Your reproduced baseline using `gpt-4o-2024-11-20` is 46.0. The 2.5pp gap is small enough to plausibly come from a combination of proxy differences + the 90 content-filtered tasks + the 7:3 split, even if both setups happen to hit the same underlying snapshot.

### Q2 — Table 1 vs Table 2 semantics

- **Table 1 "Last Epoch Success Rate" (Runtime Learning)** = success rate on the **train set** at the **last epoch** (epoch 10 in our setup). Memory is built and updated continuously during training; Table 1 reports the final-epoch number on the same train set.
- **Table 2 "Transfer Learning Success Rate"** = success rate on the **validation set**, taking the **best epoch** across all epochs. "Transfer" here means: memory is constructed on the train set during training; at each epoch we also evaluate on val to see whether the train-set-derived memory transfers to unseen tasks. Table 2 reports the best val SR observed across epochs.

So in your numbers:
- Your Table-1-equivalent = `50.7` (last-epoch train SR at epoch 7)
- Your Table-2-equivalent = `47.0` (best val SR across epochs 1–7)

### Q3 — `sim_norm_mean` / `sim_norm_std`

**Use the YAML defaults: `sim_norm_mean: 0.31`, `sim_norm_std: 0.10`.** These are the values the paper experiments used; please keep them as-is rather than substituting your locally recomputed 0.186 / 0.094.

### Q4 — convergence rate

Your observed ~+1pp/epoch on the train set is in the normal range. In our experiments, MemRL typically converges around **10 epochs** on BCB (occasionally a couple more). We use 10 epochs uniformly across all benchmarks for consistency rather than tuning per-benchmark. So +1pp/epoch averaging over ~10 epochs is consistent with our setup — what matters more is the **absolute final value**, not the per-epoch slope.

### Q5 — improvement curve shape

The training curve shape is honestly not very deterministic — there's meaningful epoch-to-epoch variance across runs (different seeds, different memory ordering effects), and we don't have a single canonical shape we'd call "expected." The roughly +1pp/epoch trend you observe is consistent with what we see on average, but the exact per-epoch trajectory can fluctuate.

I'd suggest comparing the **last-epoch absolute SR after 10 epochs** , rather than trying to match an intermediate trajectory.

---

Hope this helps — let me know if anything is unclear!

**Comment by Lu-Yang666** [comment]
Hi, thank you so much for the detailed and helpful response! This really clarified a lot of things for us.

## Clarifications on our setup

1. **Code version**: We can confirm we are on the latest code version (post `dc4d50a`), and `memory_budget_tokens` is correctly set to `0` (unlimited) in our config.

2. **sim_norm_mean / sim_norm_std**: We identified that we were using locally-computed values (0.186 / 0.094) instead of the YAML defaults (0.31 / 0.10). We have now corrected this to match your recommended values. This likely contributed to suboptimal Q-value normalization in our earlier runs.

## Updated results (Epochs 8–10)

Since our original report only covered E1–E7, here are the complete 10-epoch results:

| Epoch | Train Pass/798 | Train SR | Val Pass/342 | Val SR |
|-------|---------------|----------|-------------|--------|
| E8 | 410 | **51.4%** | 160 | **46.8%** |
| E9 | 406 | 50.9% | 155 | 45.3% |
| E10 | 400 | 50.1% | 153 | 44.7% |

**Summary**:
- Runtime Learning (Last Epoch, E10): 50.1%
- Transfer Learning (Best Val, E8): 46.8%
- Peak Train (E8): 51.4%

The E9/E10 slight decline is consistent with your note about epoch-to-epoch variance.

## Next steps

We are now re-running the full experiment with the corrected `sim_norm_mean=0.31` / `sim_norm_std=0.10` values and will report back if we have further questions after the new run completes.

We'll close this issue for now. Thanks again for the thorough response — especially the context on Table 1/2 semantics and the convergence behavior, as well as the pointers on verifying the code version and the `memory_budget_tokens` bugfix, which helped us systematically rule out potential issues. Very helpful!
